bin/thingsvision.py

```diff
@@ -0,0 +1,179 @@
+    description=textwrap.dedent(
+        """ This is the thingsvision CLI. It allows you to extract features from images using state-of-the-art neural networks for Computer Vision."""
+    ),
+)
+
+common_parser = argparse.ArgumentParser(parents=[parent_parser], add_help=False)
+common_parser.add_argument(
+    "--model_name",
+    type=str,
+    default="alexnet",
+    help="Name of the model to use for feature extraction.",
+)
+common_parser.add_argument(
+    "--source",
+    type=str,
+    default="torchvision",
+    choices=["torchvision", "keras", "timm", "custom"],
+    help="Source of the model to use for feature extraction. ",
+)
+common_parser.add_argument(
+    "--device",
+    type=str,
+    default="cpu",
+    choices=["cpu", "cuda"],
+    help="Device to use for the extractor.",
+)
+
+subparsers = parent_parser.add_subparsers(
+    title="Subcommands",
+    description="Valid subcommands are",
+    help="Additional help available is available within  each subcommand",
+    dest="command",
+)
+
+parser_model = subparsers.add_parser(
+    "show_model", description="Shows the model architecture", parents=[common_parser]
+)
+
+parser_extract = subparsers.add_parser(
+    "extract_features", description="Extract features from images", parents=[common_parser]
+)
+
+parser_extract.add_argument(
+    "--image_root",
+    type=str,
+    help="Path to directory containing images",
+    default="./images",
+)
+parser_extract.add_argument(
+    "--class_names",
+    type=str,
+    help="optional list of class names for class dataset",
+    default=None,
+)
+parser_extract.add_argument(
+    "--file_names",
+    type=str,
+    help="optional list of file names for class dataset",
+    default=None,
+)
+parser_extract.add_argument(
+    "--batch_size", type=int, default=32, help="Batch size used for feature extraction."
+)
+parser_extract.add_argument(
+    "--out_path",
+    type=str,
+    default="./features",
+    help="Path to directory where features should be stored.",
+)
+parser_extract.add_argument(
+    "--model_parameters",
+    type=str,
+    default=None,
+    help="Clip vision model has to be defined in the model_parameters dictionary",
+)
+parser_extract.add_argument(
+    "--module_name",
+    type=str,
+    default="features.10",
+    help="Name of the module to use for feature extraction.",
+)
+parser_extract.add_argument(
+    "--file_format",
+    type=str,
+    default="npy",
+    help="File format to use for storing features.",
+    choices=["npy", "hdf5", "mat", "txt"],
+)
+parent_parser.add_argument(
+    "-h",
+    "--help",
+    action="help",
+    default=argparse.SUPPRESS,
+    help="Show this help message and exit.",
+)
+parent_parser.add_argument(
+    "-v",
+    "--version",
+    action="version",
+    version="%(prog)s 2.2.8",
+    help="Show program's version number and exit.",
+)
+
+
+def main():
+    args = parent_parser.parse_args()
+    if len(sys.argv) == 1:
```

Comment:
I second this.

Reply:
Yes, this would be nice. Inherently we already restrict specific choices in the argument parser. What we would probably need is a dict or list of all available model and module names and to do the checking there. This did not seem easy out of the box, but maybe I have missed something.

Reply:
Fair enough. We can also wait for this functionality after I've resolved issue #87 and @andropar or whoever has resolved issue #89. I think config classes will make this quite easy.

Reply:
Not a big issue or anything, but I thought standard practice was for multi-word arguments in CLIs was to have the dash / hyphen as the word separator and this then is converted into underscore (and referenced from there) in argparse (this is how the argparse docs also use it). I think it's just best to stick to dashes now, i.e. `--file-names` in the call but `args['file_names']` when extracting the value, but just a small observation (not a hill I'm willing to die on or anything). 

Also, it would be good if the `--version` option could reference the version file and not be hard-coded (it's easy to imagine in one of the updates, this might be forgotten about).

Reply:
Thanks for the review and the excellent hints! I have added dashes as argument in d9b2d15797219b7be5e23ebfbe1634bebc795da8 and 36b33351b95a699738c19ce3dbbb73968829f192. The version file is now also read in from the main _version.py as can be seen from 27e52113f97025a82e0495ddfe582c64242e424c.

Reply:
> > Has this been tested with the most recent thingsvision version?
> 
> @florianmahner @andropar We should probably add a CLI HOWTO to the docs.

Yes, it is now in line with the examples from the HOWTO docs. @RLeipe could you maybe add a short cli part?

Reply:
@RLeipe could you do this in consultation with @florianmahner?